sdk/src/crypto.rs

```diff
@@ -23,21 +25,83 @@ use crate::{status_code_to_bool, sys, SyscallResult};
 /// NOTE: This only supports f1 and f3 addresses.
 pub fn verify_signature(
     signature: &Signature,
-    signer: &Address,
+    addr: &Address,
     plaintext: &[u8],
 ) -> SyscallResult<bool> {
+    use fvm_shared::address::{Payload, Protocol};
+
     let sig_type = signature.signature_type();
     let sig_bytes = signature.bytes();
-    let signer = signer.to_bytes();
+
+    match sig_type {
+        SignatureType::BLS => {
+            let sig: &[u8; BLS_SIG_LEN] = sig_bytes
+                .try_into()
+                .map_err(|_| ErrorNumber::IllegalArgument)?;
+
+            let pub_key = match addr.payload() {
+                Payload::BLS(pub_key) => *pub_key,
+                _ => return Err(ErrorNumber::IllegalArgument),
+            };
+
+            verify_bls_aggregate(sig, &[pub_key], &[plaintext])
+        }
+        SignatureType::Secp256k1 => {
+            if addr.protocol() != Protocol::Secp256k1 {
+                return Err(ErrorNumber::IllegalArgument);
+            }
+
+            let sig: &[u8; SECP_SIG_LEN] = sig_bytes
+                .try_into()
+                .map_err(|_| ErrorNumber::IllegalArgument)?;
+
+            let digest = hash_blake2b(plaintext);
+
+            let addr_recovered = {
+                let pub_key = recover_secp_public_key(&digest, sig)?;
+                Address::new_secp256k1(&pub_key).expect(
+                    "recovered secp256k1 public key should always be a valid secp256k1 address",
+                )
+            };
+
+            Ok(addr == &addr_recovered)
+        }
+    }
+}
+
+pub fn verify_bls_aggregate(
+    sig: &[u8; BLS_SIG_LEN],
+    pub_keys: &[[u8; BLS_PUB_LEN]],
+    plaintexts: &[&[u8]],
+) -> SyscallResult<bool> {
+    let num_signers = {
+        let (num_signers, num_plaintexts) = (pub_keys.len(), plaintexts.len());
+        if num_signers != num_plaintexts {
+            return Err(ErrorNumber::IllegalArgument);
+        };
+        num_signers
+            .try_into()
+            .map_err(|_| ErrorNumber::IllegalArgument)?
+    };
+
+    let plaintext_lens = plaintexts
+        .iter()
+        .map(|msg| {
+            msg.len()
+                .try_into()
+                .map_err(|_| ErrorNumber::IllegalArgument)
+        })
+        .collect::<SyscallResult<Vec<u32>>>()?;
+
+    let plaintexts_concat: Vec<u8> = plaintexts.concat();
+
     unsafe {
-        sys::crypto::verify_signature(
-            sig_type as u32,
-            sig_bytes.as_ptr(),
-            sig_bytes.len() as u32,
-            signer.as_ptr(),
-            signer.len() as u32,
-            plaintext.as_ptr(),
-            plaintext.len() as u32,
+        sys::crypto::verify_bls_aggregate(
+            num_signers,
+            sig.as_ptr() as *const u8,
+            pub_keys.as_ptr() as *const u8,
+            plaintext_lens.as_ptr() as *const u8,
+            plaintexts_concat.as_ptr() as *const u8,
```

Comment:
I don't think we should be casting in any of these cases. We should be able to pass these pointers as-is (although we may need to change the syscall?).

Reply:
Well, the `pub_keys` is more debatable. But we should definitely pass `plaintext_lens` as `*const u32`.